Test/git-issues/git-issue-2013.dfy

```diff
@@ -0,0 +1,191 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:py "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+
+method Main() {
+  TestTraits();
+  TestSiblings();
+  TestException();
+  MoreTests.Test();
+  Conveyance.Test();
+}
+
+// ----- reference types -----
+
+datatype Dt<+U> = Dt(u: U)
+
+trait Tr {
+  var y: int
+}
+
+class Cl extends Tr {
+  constructor () {
+    y := 15;
+  }
+}
+
+method Print(e: Dt<Cl>) {
+  print e.u.y, "\n";
+}
+
+method CreateDtForTrait(e: Dt<Cl>) returns (d: Dt<Tr>)
+  ensures d.u == e.u
+{
+  d := e; // upcast
+}
+
+method TestTraits() {
+  var cl := new Cl();
+  var e: Dt<Cl> := Dt(cl);
+  var d: Dt<Tr> := CreateDtForTrait(e);
+  cl.y := cl.y + 1;
+  var f: Dt<Cl> := d; // downcast
+
+  Print(f); // 16
+  Print(e); // 16
+}
+
+// ----- sibling types -----
+
+type Even = x | x % 2 == 0
+
+method TestSiblings() {
+  // downcasts
+  var a: Dt<int> := Dt(10);
+  var b: Dt<nat> := a;
+  var c: Dt<Even> := a;
+
+  // upcasts
+  b := Dt(11);
+  c := Dt(12);
+  a := b;
+  a := c;
+
+  // sideway casts
+  b := Dt(20);
+  c := b;
+  c := Dt(30);
+  b := c;
+
+  print a, " ", b, " ", c, "\n"; // Dt.Dt(12) Dt.Dt(30) Dt.Dt(30)
+}
+
+// ----- Result -----
+
+datatype Result<+T, +R> = | Success(value: T) | Failure(error: R)
+
+trait Exception {}
+
+class MyClassException extends Exception { }
+
+class MyClass {
+  const error: Exception
+
+  constructor () {
+    error := new MyClassException;
+  }
+
+  function method Foo(u: int): Result<int, Exception> {
+    if u < 8 then
+      Success(8)
+    else
+      Failure(error)
+  }
+}
+
+method TestException() {
+  var c := new MyClass();
+  print c.Foo(2), " ", c.Foo(22), "\n"; // Result.Success(8) Result.Failure(_module.MyClassException)
+}
+
+module MoreTests {
+  type Even = x | x % 2 == 0
+
+  datatype Dt<T> = Make(T)
+  datatype DtCovariant<+T> = MakeCovariant(T)
+
+  trait Z { }
+  trait X extends Z { }
+  trait Y extends Z { }
+  class C extends X, Y { }
+
+  method Test() {
+    var c := new C;
+
+    var s: set<int> := {100};
+    var t: set<nat> := s;
+    var u: set<Even> := s;
+    t := u;
+    u := t;
```

Comment:
```suggestion
    s := t;
```
Both seem like a decent idea.

Reply:
For further testing, I prefer `u`, since the types `Even` and `nat` are not subtypes of each other.

Reply:
I added your test, too (here and in a few more places).